Document Code Review Principles (#3596)
tock/tock: doc/CodeReview.md

## Comments and Review Criteria

Most pull requests will receive comments and reviews via Github from core team
members and other interested parties. Likely, only trivial pull requests (e.g.
spelling/formatting fixes, test fixes, or documentation/comment updates) will be
merged without discussion.

The types and detail of the comments will vary based on the type of change and
the location within the repository of the changes. Changes marked "significant"
will receive a more thorough review. Similarly, changes to code that is used
widely (e.g. changes within the core `kernel` crate) will be more scrutinized.

To help pull request reviewers and pull request submitters alike, we document
review principles that will be used when evaluating pull requests.

### General Review Principles

**PR Mechanics**

- Is this a self-contained change, or should portions of the pull request be
  split into separate PRs? Note, this is not evaluated by number of files or
  changed lines, but rather by the semantic meaning of the changes.
- Are the commits all relevant to the change, or are there possibly unrelated
  branches that are unintentionally included?

> - Does the provide enough explanation to help reviewers understand its purpose,

Reviewer: ```suggestion
- Does the PR provide enough explanation to help reviewers understand its purpose,
```

  and to explain the change for future readers of the code? Does the PR link to
  relevant tracking issues or other discussions? Are there existing discussions
  that should be referenced?

**Documentation and Comments**

> - Many core designs of Tock are document in specific markdown files. Does this

Reviewer: ```suggestion
- Many core designs of Tock are documented in specific markdown files. Does this
```

  interface) does it include enough helpful information for users to understand
  how to use it?
- If documentation is removed, is it clear that it is no longer accurate or
  needed? Or is there justification for removing the documentation?

**Code**

- Is `unsafe` used? If so, the changes in this PR need to be carefully
  evaluated. The purpose of `unsafe` must be documented and why it is used
  correctly must be explained in a comment.
- Is `unsafe` used when code is not actually memory or type unsafe (i.e. does
  not violate the Rust safety model)? This should not be marked `unsafe` and
  instead should likely be marked safe or use a capability.
- Does the code use interrupts and callbacks? If so, the code MUST NOT issue a
  callback from a downcall. The callback may ONLY be called in response to an
  interrupt. Using deferred calls is often necessary to remedy this.
- Are Rust features (i.e. conditional compilation and `#[cfg]`) used? These must
  be clearly motivated and documented, and are only permitted in specific cases.
- Are `lib.rs` or `mod.rs` files added? In general these should only be used to
  reference other modules and setup exports. Actual OS logic should be in

>   descriptively named files.

Reviewer: I know that we've been generally following this rule, but Rust's module system sometimes makes it very awkward to require `mod.rs` at `some::path::in::the::module::hierarchy`, but then require an extra `::spurious_module_path` at the end, just because we add another descriptively named file. In many ways we can work around this through reexports, but we also have counterexamples in the repository.

The current text says "in general these should", and that seems like fine language to me, if it's indeed not actually meant to be a hard rule.

Author: We have a legacy dating back to the beginnings of tock of using lib.rs in arch crates which I guess is ok, hence my hedging. (Although, actually looking at the commits my excuse isn't really valid, we _had_ a different file and moved it to lib.rs :( https://github.com/tock/tock/commit/2c1d6206d91f57a5d402502617e42b0222ffcdc0)

That being said, for me it is a rule: either you have little code/modularity and do not need a crate/folder, or you have enough code/modularity to warrant a crate/folder and the structure should plan for mode code to be added in the future (i.e. do not use the mod/lib file).

- `static_init!()` (and similar) must only be called from board crates.
- Is any new functionality both publicly exported and harmful to overall safety
  (not just correctness) if called at the wrong time? If so, this should likely

>   be guarded with a capability.

Reviewer: ```suggestion
  be guarded with a capability or `unsafe`.
```

Given this mentions safety. We'd want to evaluate on a per-case basis whether "safety" means functional safety, or Rust memory safety.

Author: If rust thinks something is unsafe then the unsafe keyword will be there and a previous principle applies.

This case is only concerning non-`unsafe` safety (e.g. running the kernel loop). I don't think it is possible to use safe apis to create unsafe code that needs to be marked unsafe. But perhaps there is a way to word this to reduce confusion.

> - Is any new functionality both publicly exported and harmful to overall safety
>   (not just correctness) if called at the wrong time? If so, this should likely
>   be guarded with a capability.

Reviewer 2: ```suggestion
- Is any new functionality both publicly exported and have invariants which
  cannot be enforced by the type system or other automated means (e.g., they
  provide access to sensitive core kernel data structures). If so, this should likely
  be guarded with a capability.
```

Maybe this?

> - Uses of `#inline` directives should be explained why they are needed.

Reviewer 3: ```suggestion
- Uses of `#inline` directives should explain why they are needed.
```

> - Uses of `#inline` directives should explain why they are needed.

Reviewer 2: Should this generalize to any compiler directives (excepting `#[test]`)?

Should this maybe be more specific about where/how the explanation exists, i.e.
> should explain in an adjacent comment why they are needed.

comments to explain their purpose, and all functions and data structures should
be clearly documented with `///` comments. Often additional documentation in
discrete markdown files is required as well.

Additionally, particularly subtle or extensively discussed rationale should be
included in the source file directly (often with a `//` comment). This leaves a
clear trace of how key design decisions in Tock were decided and why certain
aspects may not use the most intuitive design. This helps avoid re-hashing
discussions and assist new users with understanding the kernel.

All `unsafe` usage MUST be accompanied by a comment starting with `### Safety`
that discusses exactly why the unsafe code is necessary and what checks are
needed and completed to ensure the use of `unsafe` does not trigger undefined
behavior.

All new exports from the core kernel crate must be carefully examined. Certain
functionality is only safe within the core kernel. As essentially every crate in
Tock uses `kernel` as a dependency, anything exported can be used broadly.
Functionality which is sensitive but _must_ be exported must be guarded by a
capability.

#### HILs

> New HILs should follow the TRD on HIL design.

Author: I'll add this, but it is frustrating that our scheme explicitly breaks permalinking.

> New HILs should follow the TRD on HIL design.

Reviewer 3: Presumably it would say "obsoleted". This just happened to me, as I opened trd2 first, and I figured it out.

HILs should be well documented and not specifically matched to a single hardware
platform.

All valid errors should be enumerated.

HIL naming should be reasonably consistent and clear.

#### Capsules

Capsules should explain what they do in comments but do not need to be
rigorously commented.

**Virtualizers**

Virtualizers multiplex an underlying resource for multiple users.

> - The `Mux` struct should handle all interrupts, and route callback to specific

Reviewer 3: ```suggestion
- The `Mux` struct should handle all interrupts, and route callbacks to specific
```

  virtualizer users.
- The virtualizer should provide the same interface (i.e. HIL) as it uses from
  the underlying shared resource.

**Syscall Drivers**

Syscall drivers implement `SyscallDriver` to provide interfaces for userspace.

- These drivers must support multiple processes. They do not need to be fully

>   virtualized, but they must not break if multiple processes use them.

Reviewer 3: I think this needs more clarity: Is it okay for them to be first-come-first-served and reject later processes? (but not crash or do something invalid)

Author: Yes that is what that was trying to say.

> - These drivers must support multiple processes. They do not need to be fully
>   virtualized, but they must not break if multiple processes use them.

Reviewer 2: ```suggestion
- These drivers must support potential calls from multiple processes. They do
   not need to be fully virtualized, e.g. a driver which rejects syscalls from all
   but the first process to access it is acceptable, but drivers must not break if
   multiple processes attempt access.
```

Maybe this?

- They must return `CommandReturn::SUCCESS` for `command_id==0`.
- They should use the first argument to any upcalls as a ReturnCode.
- They should only provide an interface to userspace on top of some resource,
  and should not implement additional functionality which may also be useful
  within the kernel. The additional functionality should be a separate capsule.

#### Chips

Often changes within chip crates are difficult to test as not many reviewers may
have the specific hardware. Review comments often rely on visual inspection of
the code.

Files in a chip crate should avoid giving the impression of functionality which
is not actually implemented. This means avoiding peripheral files which only

> contain registers, or returning `ErrorCode::NOSUPPORT`.

Reviewer 2: ```suggestion
contain registers or return `ErrorCode::NOSUPPORT` for all methods. A peripheral
must implement at least basic functionality to be merged in mainline Tock.
```

Chip crates should be properly named. Many chips use nested crates to represent
families of chips and to share implementations.

Rust `cfg` features should be avoided. However, in circumstances where different
chips differ in very small ways, and those differences are well understand and
likely documented in a datashet, chip-variant configs may be used. They should
be contained to a single file (i.e. not scattered throughout the crate). It

> should be entirely unambiguous whether a feature is set or not (i.e. it should
> be based on physical hardware where it is obvious which chip a user has).

Reviewer 2: ```suggestion
should be entirely unambiguous whether a feature is set or not (i.e. it should
be based on physical hardware where it is obvious which chip a user has).
Generally, this means `cfg` directives should be an explicit list of chips or'd
together. Rarely, if ever, is a `cfg(not ...)` the correct approach for anything
outside of unit tests.
```
